cylc/flow/data_store_mgr.py

```diff
@@ -539,19 +542,173 @@ def generate_definition_elements(self):
         self.descendants = descendants
         self.parents = parents
 
-    def generate_ghost_task(self, t_id, tp_id, point_string):
+    def increment_graph_window(
+            self, name, point, flow_label,
+            edge_distance=0, active_id=None, descendant=False):
+        """Generate graph window about given origin to n-edge-distance.
+
+        Args:
+            name (str):
+                Task name.
+            point (cylc.flow.cycling.PointBase):
+                PointBase derived object.
+            flow_label (str):
+                Flow label used to distinguish multiple runs.
+            edge_distance (int):
+                Graph distance from active/origin node.
+            active_id (str):
+                Active/origin node id.
+            descendent (bool):
+                Is the current node a direct descendent of the active/origin.
+
+        Returns:
+
+            None
+
+        """
+        # Create this source node
+        s_node = TaskID.get(name, point)
+        s_id = f'{self.workflow_id}{ID_DELIM}{point}{ID_DELIM}{name}'
+        if active_id is None:
+            active_id = s_id
+
+        # Setup and check if active node is another's boundary node
+        # to flag it's paths for pruning.
+        if edge_distance == 0:
+            self.n_window_edges[active_id] = set()
+            self.n_window_boundary_nodes[active_id] = {}
+            self.n_window_nodes[active_id] = set()
+            if active_id in self.prune_trigger_nodes:
+                self.prune_flagged_nodes.update(
+                    self.prune_trigger_nodes[active_id])
+                del self.prune_trigger_nodes[active_id]
+                self.prune_pending = True
+
+        # This part is vital to constructing a set of boundary nodes
+        # associated with the current Active node.
+        graph_children = generate_graph_children(
+            self.schd.config.get_taskdef(name), point)
+        if (
+                edge_distance > self.n_edge_distance or
+                not any(graph_children.values())
+        ):
+            if descendant or edge_distance == 0:
+                self.n_window_boundary_nodes[
+                    active_id].setdefault(edge_distance, set()).add(s_id)
+
+        if edge_distance > self.n_edge_distance:
+            return
+
+        self.n_window_nodes[active_id].add(s_id)
+        # Generate task node
+        self.generate_ghost_task(s_id, name, point, flow_label)
+
+        edge_distance += 1
+
+        # TODO: xtrigger is suite_state edges too
+        # Reference set for workflow relations
+        for _, items in graph_children.items():
+            if edge_distance == 1:
+                descendant = True
+            self._expand_graph_window(
+                s_id, s_node, items, active_id, flow_label, edge_distance,
+                descendant, False)
+
+        for _, items in generate_graph_parents(
+                self.schd.config.get_taskdef(name), point).items():
```

Comment:
I think these two `for` loops don't need to iterate `.items()` if they are discarding the key? Could be just `for items in ().values()` ? Probably doesn't make much difference in performance.

Reply:
True, I dashed them out temporarily, thanks for the reminder.